base/threadingconstructs.jl

```diff
@@ -22,83 +22,67 @@ See also: `BLAS.get_num_threads` and `BLAS.set_num_threads` in the
-                tid = 1
-                len, rem = lenr, 0
+            local range_len = length(range)
+            range_len == 0 && return
+            local ntasks = if $(schedule === :dynamic)
+                nthreads()
+            elseif threadid() != 1 || ccall(:jl_in_threaded_region, Cint, ()) != 0
+                $(if schedule === :static
+                    :(error("`@threads :static` can only be used from thread 1 and not nested"))
+                else
+                    # only use threads when called from thread 1, outside @threads :static
+                    1
+                end)
             else
-                tid = threadid()
-                len, rem = divrem(lenr, nthreads())
+                nthreads()
             end
-            # not enough iterations for all the threads?
-            if len == 0
-                if tid > rem
-                    return
+            local n_per_part = div(range_len, ntasks, RoundUp)
+            if ntasks == 1
+                # run locally as a simple loop
+                for $(esc(lidx)) in range
+                    $(esc(lbody))
                 end
-                len, rem = 1, 0
-            end
-            # compute this thread's iterations
-            f = firstindex(r) + ((tid-1) * len)
-            l = f + len - 1
-            # distribute remaining iterations evenly
-            if rem > 0
-                if tid <= rem
-                    f = f + (tid-1)
-                    l = l + tid
-                else
-                    f = f + rem
-                    l = l + rem
+            else
+                local parts = Iterators.partition(range, n_per_part)
+                local nparts = length(parts)
+                $(if schedule !== :dynamic
+                    :(ccall(:jl_enter_threaded_region, Cvoid, ()))
+                end)
+                try
+                    local tasks = Vector{Task}(undef, nparts)
+                    local tid = 0
+                    for part in parts
+                        local t = @task begin
+                            local $(esc(lidx))
+                            for $(esc(lidx)) in part
+                                $(esc(lbody))
+                            end
+                        end
+                        $(if schedule === :dynamic
+                            :(t.sticky = false)
+                        else
+                            :(t.sticky = true)
+                            :(ccall(:jl_set_task_tid, Cint, (Any, Cint), t, tid))
+                        end)
+                        tasks[tid + 1] = t
+                        schedule(t)
+                        tid += 1
+                    end
+                    for i = 1:nparts
+                        wait(tasks[i])
+                    end
+                finally
+                    $(if schedule !== :dynamic
+                        :(ccall(:jl_exit_threaded_region, Cvoid, ()))
+                    end)
                 end
```

Comment:
Hmm... Wouldn't it be cleaner to do less in the macro? I'd simply modify the old `threading_run` to create a closure like this

```julia
function threading_run(fun, static)
    static && ccall(:jl_enter_threaded_region, Cvoid, ())
    n = nthreads()
    tasks = Vector{Task}(undef, n)
    for i = 1:n
        t = Task(() -> fun(i)) # pass in tid
        t.sticky = static
        static && ccall(:jl_set_task_tid, Cint, (Any, Cint), t, i-1)
        tasks[i] = t
        schedule(t)
    end
    try
        for i = 1:n
            wait(tasks[i])
        end
    finally
        static && ccall(:jl_exit_threaded_region, Cvoid, ())
    end
end
```

And then let `threadsfor_fun` take a `tid` instead of looking it up via `threadid`:

```julia
function threadsfor_fun(tid=1, onethread=false)
```

Not only that it has the benefit of making the patch and the code small but also that it actually distributes the iteration space a bit better than `Iterators.partition`. For example, if I have 4 threads and 6 items, old `threadsfor_fun` distributes the items as `[2, 2, 1, 1]` but with `Iterators.partition` we have `[2, 2, 2, 0]`. If the workload is skewed and the last two items have much larger run-time, the performance of `Iterators.partition`-based approach can be noticeably bad. This is presumably why `threadsfor_fun` tries to distribute items evenly.

Reply:
Great, thanks. I also just noticed the `Iterators.partition` also has more execution overhead. I'll implement your suggestion